### ISSUE TITLE
Household and Individual IDs

### ISSUE THREAD

**Issue by buscandoaverroes** [worldbank/gld#33]
Since I'm not provided with how many HH are (the PSA and ILO do provide numbers for how many individual cases there are, but not households), I can't find a way to confirm how many Households there are in each survey year. My own variable coding has resulted in some years with drastically fewer households than others, which means I should review the variable coding and ensure that I'm extra confident in the number of households that it produces.

**Comment by gronert-m** [comment]
@buscandoaverroes what does "drastically fewer households" mean in this context. If the number of individuals is as expected and similar to other years but the number of households varies, then the household size is unstable over time. That indeed would be odd.

If it is that the number of observations as a whole falls, then it is normal that households are fewer. But the question would be why is the number of observations smaller. Am I understanding this correctly?

**Comment by gronert-m** [comment]
Thanks for explaining that. I wonder if that coefficient is stable across the quarters. Let me know if you can't find the reason for this and we can do some more forensics.

**Comment by buscandoaverroes** [comment]
After a bit more digging, some (re)discoveries:

- not every year has a pre-loaded household id number, for example, 2018, and many earlier years. 
- The earlier years have an `hcn` or household control number variable, but this has only about [~6,000 observations](https://www.ilo.org/surveyLib/index.php/catalog/2012/data-dictionary/F10?file_name=PHL_2002_LFS_Q1_v01_M_SPSS) per quarter and isn't enough to uniquely identify observations along with the preloaded `line number` variable.

Because of this second point, I mistakenly thought that I needed to construct the household identifier variable in cases where it wasn't provided.

The way my code works is I just take all non-missing demographic variables across all observations, like region, stratum, etc, to make a "household" id. Then I sort by `_n` within each "family". I check this at the end by `isid` so I know that, at least, `hhid` and `pid` together uniquely identify the observations. I don't use the `line number` variable, but in hindsight I problably should have. By definition, my constructed "hhid" along with "_n" -- so long as it is specific enough --  will produce a "unique id" but it may not actually be a valid indicator of what the Household unit is. 

This explains the variation in number of Households across survey years: those in which I'm provided a HHID are fairly constant, but those where I have to construct it are sort of all over the place.

The resulting questions is then: what should I do in cases where I'm not given a household id? I think we can talk about this in our meeting.

**Comment by gronert-m** [comment]
Yeah, we can talk about it. I was just looking at April and July 2018 and in both cases it would seem that `pufotherid` (terrible name, other of what?) is a household ID and the `line number` (`pufc01_lno`) is unique. `pufotherid` gives then an average household size (individual obs / # of HHs) of 4.53 in July, 4.39 in April. In January it was 4.36 using `pufhhnum` and `pufc01_lno`. I think it would make sense to use `line number` as you said.

As you say, we can discuss this later today.

**Comment by buscandoaverroes** [comment]
Ah, well that would do it then -- Thanks for looking into this!

**Comment by buscandoaverroes** [comment]
In revisiting this, I'm noticing that:

- looking in depth in just one round (April 2018, the line number variable seems to be accurate)
- Years 2004 and later have unique HHID, usually called `hhnum` or `pufhhnum`
- in years prior to 2004, a set of variables uniquely identifies what appears to be the household, but this set changes from survey to survey. This gives the expected number of household groupings in each survey and does not change when adding additional demographic variables.
- Even when the set of household identifying variables is identified, there are still sometimes a few duplicate observations across most or all variables, which should be cleaned.

**Comment by buscandoaverroes** [comment]
For 2003, there's not a common set of variables present in all 4 rounds that uniquely identify households.

**Comment by buscandoaverroes** [comment]
Update: For all years prior to 2008, at least one round lacks a HHID variable, so that will need to be constructed for those rounds

**Comment by buscandoaverroes** [comment]
Additional update: when creating the HHID variable, I'm noticing that a few rounds have about a dozen or fewer entries where there are obvious duplicate observations, which need to be dealt with before making the unique HHID. For the time being, I'm dropping the observations that are clearly duplicates (such as duplicated on most or all variables), leaving one observation in the duplicate set in the data.

**Comment by buscandoaverroes** [comment]
Also, I've run into a slight logistical issue: if you have a year where you need two distinct groups of variables to determine a unique HHID (2003, 2004, 2006, 2007), it's quite difficult they way the current code is set up. The general order of things is 
```
append/harmonize variable names
create HHIDs
Labor and Edu modules
```
and this is great, but the code for generating the HHIDs assumes that there's a single dataset in memory, and it becomes clumsy quite quickly if you have to conditionally generate the HHID (ie making temp files, saving external files etc., or simply creating the HHID variable become appending). I have a feeling that the actual aligning of the dataset will present further challenges so I'm going to "table" this and essentially address the aligning and HHID creating for these four years at once.

**Comment by buscandoaverroes** [comment]
Also noting that in subsequent years (such as 2017) where there is supposedly a unique HHID given , that HHID does not uniquely identify observations with the line number and round. There are many duplicates -- thousands -- to the point where I may simply construct my own for these cases

**Comment by buscandoaverroes** [comment]
resolved in #56
